From fdf575828e66146aa15aa86936afc78664583c7a Mon Sep 17 00:00:00 2001
From: Doug Ledford <dledford@redhat.com>
Date: Wed, 24 Sep 2014 17:28:11 -0400
Subject: [infiniband] ipoib: Consolidate rtnl_lock tasks in workqueue

Message-id: <fa087a08eb34c0afcb2ddf5b126b4d0c57ea0a35.1411578494.git.dledford@redhat.com>
Patchwork-id: 96672
O-Subject: [Patch RHEL7 18/26] IPoIB: Consolidate rtnl_lock tasks in workqueue
Bugzilla: 1107621
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

Setting the mtu can safely be moved to the carrier_on_task, which keeps
us from needing to take the rtnl lock in the join_finish section.

Signed-off-by: Doug Ledford <dledford@redhat.com>
(cherry picked from commit 9abc6963c09ff1b364a8552853de30502d6603cc)
Signed-off-by: Doug Ledford <dledford@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/ulp/ipoib/ipoib_multicast.c b/drivers/infiniband/ulp/ipoib/ipoib_multicast.c
index d4e0057..a0a4285 100644
--- a/drivers/infiniband/ulp/ipoib/ipoib_multicast.c
+++ b/drivers/infiniband/ulp/ipoib/ipoib_multicast.c
@@ -190,12 +190,6 @@ static int ipoib_mcast_join_finish(struct ipoib_mcast *mcast,
   spin_unlock_irq(&priv->lock);
   priv->tx_wr.wr.ud.remote_qkey = priv->qkey;
   set_qkey = 1;
-
-  if (!ipoib_cm_admin_enabled(dev)) {
-   rtnl_lock();
-   dev_set_mtu(dev, min(priv->mcast_mtu, priv->admin_mtu));
-   rtnl_unlock();
-  }
  }
 
  if (!test_bit(IPOIB_MCAST_FLAG_SENDONLY, &mcast->flags)) {
@@ -371,6 +365,8 @@ void ipoib_mcast_carrier_on_task(struct work_struct *work)
  }
 
  rtnl_lock();
+ if (!ipoib_cm_admin_enabled(priv->dev))
+  dev_set_mtu(priv->dev, min(priv->mcast_mtu, priv->admin_mtu));
  netif_carrier_on(priv->dev);
  rtnl_unlock();
 }
-- 
1.7.1